tsl/src/dist_backup.c

```diff
@@ -0,0 +1,194 @@
+	HeapTuple tuple;
+
+	tupdesc = BlessTupleDesc(tupdesc);
+	if (node_name == NULL)
+	{
+		nulls[AttrNumberGetAttrOffset(Anum_restore_point_node_name)] = true;
+		values[AttrNumberGetAttrOffset(Anum_restore_point_node_type)] =
+			CStringGetTextDatum(TS_ACCESS_NODE_TYPE);
+	}
+	else
+	{
+		values[AttrNumberGetAttrOffset(Anum_restore_point_node_name)] = CStringGetDatum(node_name);
+		values[AttrNumberGetAttrOffset(Anum_restore_point_node_type)] =
+			CStringGetTextDatum(TS_DATA_NODE_TYPE);
+	}
+	values[AttrNumberGetAttrOffset(Anum_restore_point_lsn)] = LSNGetDatum(lsn);
+	tuple = heap_form_tuple(tupdesc, values, nulls);
+	return HeapTupleGetDatum(tuple);
+}
+
+Datum
+create_distributed_restore_point(PG_FUNCTION_ARGS)
+{
+	const char *name = TextDatumGetCString(PG_GETARG_DATUM(0));
+	DistCmdResult *result_cmd;
+	FuncCallContext *funcctx;
+	XLogRecPtr lsn;
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		int name_len = strlen(name);
+		MemoryContext oldctx;
+		TupleDesc tupdesc;
+		char *sql;
+
+		if (name_len >= MAXFNAMELEN)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("restore point name is too long"),
+					 errdetail("Maximum length is %d, while provided name has %d chars.",
+							   MAXFNAMELEN - 1,
+							   name_len)));
+
+		if (RecoveryInProgress())
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 (errmsg("recovery is in progress"),
+					  errdetail("WAL control functions cannot be executed during recovery."))));
+
+		if (!XLogIsNeeded())
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 errmsg("WAL level '%s' is not sufficient for creating a restore point",
+							GetConfigOptionByName("wal_level", NULL, false)),
+					 errhint("Set wal_level must to \"replica\" or \"logical\" at server start.")));
+
+		if (!superuser())
+			ereport(ERROR,
+					(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+					 errmsg("must be superuser to create restore point")));
+
+		if (!ts_guc_enable_2pc)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_OPERATION_NOT_SUPPORTED),
+					 errmsg("two-phase commit transactions are not enabled"),
+					 errhint("Set timescaledb.enable_2pc to TRUE.")));
+
+		if (dist_util_membership() != DIST_MEMBER_ACCESS_NODE)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_OPERATION_NOT_SUPPORTED),
+					 errmsg("distributed restore point must be created on the access node"),
+					 errhint("Connect to the access node and create the distributed restore point "
+							 "from there.")));
+
+		/*
+		 * In order to achieve synchronization across the multinode cluster,
+		 * we must ensure that the restore point created on the access node is
+		 * synchronized with each data node.
+		 *
+		 * We must avoid possible situation when a concurrent remote transaction
+		 * commits on the access node during the creation of the distributed
+		 * restore point and gets into the data nodes WAL before the
+		 * restore point marker written.
```

Comment:
I found the comment a bit hard to follow. Suggest something like:
```suggestion
		 * We must ensure that no concurrent prepared transactions are 
		 * committed (COMMIT PREPARED) while we create the restore point.
		 * Otherwise, the distributed restore point might include prepared transactions
		 * that have committed on some data nodes but not others, leading to an
		 * inconsistent state when the distributed database is restored from a backup
		 * using the restore point.
```